gtsfm/data_association/data_assoc.py

```diff
@@ -196,6 +202,44 @@ def run(
 
         return connected_data, data_assoc_metrics
 
+    def run_triangulation(
+        self,
+        cameras: Dict[int, gtsfm_types.CAMERA_TYPE],
+        tracks_2d: List[SfmTrack2d],
+    ) -> Tuple[List[Delayed], List[Delayed], List[Delayed]]:
+        """Performs triangulation of 2D tracks in parallel.
+
+        Ref: https://docs.dask.org/en/stable/delayed-best-practices.html#compute-on-lots-of-computation-at-once
+
+        Args:
+            cameras: list of cameras wrapped up as Delayed.
+            tracks_2d: list of tracks wrapped up as Delayed.
+
+        Returns:
+            sfm_tracks: List of triangulated tracks.
+            avg_track_repoj_errors: List of average reprojection errors per track.
+            triangulation_exit_codes: exit codes for each triangulation call.
+        """
+        # Initialize 3D landmark for each track
+        point3d_initializer = Point3dInitializer(cameras, self.triangulation_options)
+
+        # Loop through tracks and and generate delayed triangulation tasks.
+        triangulation_results = []
+        for track_2d in tracks_2d:
+            triangulation_results.append(dask.delayed(point3d_initializer.triangulate)(track_2d))
```

Comment:
can we do an nout=3 in dask.delayed, and fill in sfm_tracks, avg_track_reproj_errors, triangulation_exit_codes here instead of one more loop?

Reply:
Surprisingly, separating the variables before calling `compute` is slower. The following code took 1.17 minutes:

```python
def run_triangulation(
        self,
        cameras: Dict[int, gtsfm_types.CAMERA_TYPE],
        tracks_2d: List[SfmTrack2d],
    ) -> Tuple[List[Delayed], List[Delayed], List[Delayed]]:
        """Performs triangulation of 2D tracks in parallel.

        Ref: https://docs.dask.org/en/stable/delayed-best-practices.html#compute-on-lots-of-computation-at-once

        Args:
            cameras: list of cameras wrapped up as Delayed.
            tracks_2d: list of tracks wrapped up as Delayed.

        Returns:
            sfm_tracks: List of triangulated tracks.
            avg_track_repoj_errors: List of average reprojection errors per track.
            triangulation_exit_codes: exit codes for each triangulation call.
        """
        # Initialize 3D landmark for each track
        point3d_initializer = Point3dInitializer(cameras, self.triangulation_options)

        # Loop through tracks and and generate delayed triangulation tasks.
        delayed_sfm_tracks, delayed_avg_track_reproj_errors, delayed_triangulation_exit_codes = [], [], []
        for track_2d in tracks_2d:
            delayed_sfm_track, delayed_avg_track_reproj_error, delayed_triangulation_exit_code = dask.delayed(
                point3d_initializer.triangulate, nout=3
            )(track_2d)
            delayed_sfm_tracks.append(delayed_sfm_track)
            delayed_avg_track_reproj_errors.append(delayed_avg_track_reproj_error)
            delayed_triangulation_exit_codes.append(delayed_triangulation_exit_code)

        # Perform triangulation in parallel.
        sfm_tracks, avg_track_reproj_errors, triangulation_exit_codes = dask.compute(
            delayed_sfm_tracks, delayed_avg_track_reproj_errors, delayed_triangulation_exit_codes
        )

        return sfm_tracks, avg_track_reproj_errors, triangulation_exit_codes
```

Maybe the scheduler isn't intelligent enough to realize the three inputs are from the same function call?